server/service/src/rnr_form/insert.rs

```diff
@@ -0,0 +1,160 @@
+};
+
+use super::{
+    query::get_rnr_form,
+    validate::{
+        check_period_exists, check_program_exists, check_rnr_form_does_not_exist,
+        check_rnr_form_exists_for_period,
+    },
+};
+#[derive(Default, Debug, PartialEq, Clone)]
+pub struct InsertRnRForm {
+    pub id: String,
+    pub supplier_id: String,
+    pub program_id: String,
+    pub period_id: String,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum InsertRnRFormError {
+    DatabaseError(RepositoryError),
+    InternalError(String),
+    RnRFormAlreadyExists,
+    SupplierDoesNotExist,
+    SupplierNotVisible,
+    NotASupplier,
+    ProgramDoesNotExist,
+    PeriodDoesNotExist,
+    PeriodNotInProgramSchedule,
+    RnRFormAlreadyExistsForPeriod,
+    NewlyCreatedRnRFormDoesNotExist,
+}
+
+pub fn insert_rnr_form(
+    ctx: &ServiceContext,
+    input: InsertRnRForm,
+) -> Result<RnRForm, InsertRnRFormError> {
+    let rnr_form = ctx
+        .connection
+        .transaction_sync(|connection| {
+            validate(connection, &ctx.store_id, &input)?;
+
+            let rnr_form = generate(input, &ctx.store_id);
+
+            let rnr_form_repo = RnRFormRowRepository::new(connection);
+
+            rnr_form_repo.upsert_one(&rnr_form)?;
+
+            activity_log_entry(
+                ctx,
+                ActivityLogType::RnrFormCreated,
+                Some(rnr_form.id.clone()),
+                None,
+                None,
+            )?;
+
+            get_rnr_form(ctx, rnr_form.id)
+                .map_err(InsertRnRFormError::DatabaseError)?
+                .ok_or(InsertRnRFormError::NewlyCreatedRnRFormDoesNotExist)
+        })
+        .map_err(|err| err.to_inner_error())?;
+
+    Ok(rnr_form)
+}
+
+fn validate(
+    connection: &StorageConnection,
+    store_id: &str,
+    input: &InsertRnRForm,
+) -> Result<(), InsertRnRFormError> {
+    if !check_rnr_form_does_not_exist(connection, &input.id)? {
+        return Err(InsertRnRFormError::RnRFormAlreadyExists);
+    }
+
+    check_other_party(
+        connection,
+        store_id,
+        &input.supplier_id,
+        CheckOtherPartyType::Supplier,
+    )?;
+
+    if check_program_exists(connection, &input.program_id)?.is_none() {
+        return Err(InsertRnRFormError::ProgramDoesNotExist);
+    }
+
+    let period = match check_period_exists(connection, &input.period_id)? {
+        Some(period) => period,
+        None => {
+            return Err(InsertRnRFormError::PeriodDoesNotExist);
+        }
+    };
+
+    let period_schedule_ids = ProgramRequisitionSettingsRowRepository::new(connection)
+        .find_many_by_program_id(&input.program_id)?
+        .iter()
+        .map(|s| s.period_schedule_id.clone())
+        .collect::<Vec<String>>();
+
+    if !period_schedule_ids.contains(&period.period_schedule_id) {
+        return Err(InsertRnRFormError::PeriodNotInProgramSchedule);
+    }
+
+    if check_rnr_form_exists_for_period(connection, &input.period_id, &input.program_id)?.is_some()
```

Comment:
Realising I should probably also check that this is next period in the sequence here right? Prevent skipping of periods...

And maybe also that its not a period in the future?

Reply:
I guess if we do allow RnR forms to be skipped, we'd want to do some business logic, such as mark the old ones as skipped or something?
Keen to hear Mark's thoughts on that one. As far as the ledger & stock usage goes how bad is skipping?
I'm thinking potentially you could come to store that hasn't submitted and RnR form for a year, I assume in that scenario you'd probably want to just skip the old ones and start fresh somehow?

Reply:
Yeah talking with Mark and Craig, we won't allow skipping. Sounds like if it has been a really long time, they'd possibly start a new period schedule (i.e. 2024 schedule with monthly periods) ... but if not, just going through and generating r&rs with the pre-populated data wouldn't take too long, then can make adjustments in the current one